internal/nas/handler.go

```diff
@@ -10,39 +10,49 @@ import (
 	"github.com/free5gc/nas"
 )
 
-func HandleNAS(ue *amf_context.RanUe, procedureCode int64, nasPdu []byte, initialMessage bool) {
+func HandleNAS(ranUe *amf_context.RanUe, procedureCode int64, nasPdu []byte, initialMessage bool) {
 	amfSelf := amf_context.GetSelf()
 
-	if ue == nil {
+	if ranUe == nil {
 		logger.NasLog.Error("RanUe is nil")
 		return
 	}
 
 	if nasPdu == nil {
-		ue.Log.Error("nasPdu is nil")
+		ranUe.Log.Error("nasPdu is nil")
 		return
 	}
 
-	if ue.AmfUe == nil {
-		ue.AmfUe = amfSelf.NewAmfUe("")
-		gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ue.AmfUe, ue)
+	if ranUe.AmfUe == nil {
+		if ranUe.FindAmfUe != nil && !ranUe.FindAmfUe.CmConnect(ranUe.Ran.AnType) {
+			// models.CmState_IDLE
+			gmm_common.ClearHoldingRanUe(ranUe.FindAmfUe.RanUe[ranUe.Ran.AnType])
+
+			ranUe.AmfUe = ranUe.FindAmfUe
+			gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ranUe.AmfUe, ranUe)
+			ranUe.FindAmfUe = nil
+		} else {
+			// New AmfUe
+			ranUe.AmfUe = amfSelf.NewAmfUe("")
+			gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ranUe.AmfUe, ranUe)
+		}
 	}
 
-	msg, integrityProtected, err := nas_security.Decode(ue.AmfUe, ue.Ran.AnType, nasPdu, initialMessage)
+	msg, integrityProtected, err := nas_security.Decode(ranUe.AmfUe, ranUe.Ran.AnType, nasPdu, initialMessage)
 	if err != nil {
-		ue.AmfUe.NASLog.Errorln(err)
+		ranUe.AmfUe.NASLog.Errorln(err)
```

Comment:
Assume that we have an existing UE context, which is in CM-CONNECTED state.
We will release it if the new UE context has a valid security context in line 53.